### ISSUE TITLE
Testing deployd extensions in a deployd-instance

### ISSUE THREAD

**Issue by simpleTechs** [deployd/deployd#506]
As we are writing a lot of extensions fro deployd lately and we want to maintain them across a developing core, we'd love to integrate the unit-tests we have into the modules, so that they can be run automatically.

What we do right now is rather crude: We have a single deployd environment that hosts all the tests for our extensions and we use it to run them against a new version of deployd. This comes with two issues: We cannot use travis on our modules and we cannot publish tests, as they'd be somewhat un-runnable without a great deal of work.

To change that, I think we need some way of require'ing deployd as a dev-dependency and injecting our module into this deployd instance.
Then we'd want to be able to run tests (static, as well as integration tests) inside this instance.

This issue is merely a call for help, as we can't seem to get this working with the current version. Maybe anyone of you has a great idea on how to accomplish this.
I'm not 100% sure on this, but from what I get, #476 would've been a step into this direction, wouldn't it?

Thanks in advance!

**Comment by NicolasRitouet** [comment]
Testing is indeed a big topic, I'd love to see some improvements in this direction.
Once we're done with the next release, we'll try to gather ideas and brainstorm to suitable solution for everybody (module developers, contributors and end users)

**Comment by 0ff** [comment]
Just as a quick follow-up to this issue, some time ago I went with @andreialecu's changes and added a test-runner.js which could be included in module-tests. This is not the "final version", but rather a draft. I would imagine that a module developer would include some file into his tests, so that deployd will be made "ready" (i.e. running with phantomjs).

You can checkout a sample module here: [simpletechs/dpd-extension-test](https://bitbucket.org/simpletechs/dpd-extension-test).

My branch [feature/extensionTestrunner](https://bitbucket.org/simpletechs/deployd/branch/feature/extensionTestrunner) does have issues such as the dashboard being broken in tests, but for merely testing a module it seems to work great.

What do you think, is that an approach that could be integrated in deployd? (Not necessarily the way I did it, but this was easier than writing everything down :))

**Comment by andreialecu** [comment]
The way I imagine it, we could make a `deployd-dev` module, which could be installed as a dev-dependency and provide a bin that can bootstrap/run said tests.

**Comment by Maddis1337** [comment]
+1

**Comment by andreialecu** [comment]
@0ff @Maddis1337 can you take a look at https://github.com/andreialecu/deployd-test-sample/ ?

Clone it, run `npm install` then `npm test` and see if everything works. 

You can put your tests in `public/test/` and make sure they're included in index.html.

If this is sufficient, we can then release this as a yeoman generator or something.

**Comment by Maddis1337** [comment]
@andreialecu  
I am currently unable to run ```npm install``` because it fails with 

``` npm ERR! git clone https://github.com/Automattic/engine.io-client Cloning into bare repository '/Users/maddis/.npm/_git-remotes/https-github-com-Automattic-engine-io-client-aeece76f'...
npm ERR! git clone https://github.com/Automattic/engine.io-client fatal: unable to access 'https://github.com/Automattic/engine.io-client/': Server aborted the SSL handshake ```

But I read through your code, it might work, but does not solve problems with extension testing. A simple use case is that you might want to create a new ```Resource``` and test it. This will break because deployd checks the ```package.json``` for resources and you cannot have your own package as a dependency. So you the resource name in the corresponding config.json of the resource will result in a failure, because deployd does not know it. I currently avoided this by creating a new directory testapp/ which contains a whole deployd application, but I needed to have specific server run scripts.

**Comment by andreialecu** [comment]
@Maddis1337 I think you misunderstood how it's supposed to be used.

Simply clone the repo somewhere, such as in `testapp`, then run `cd testapp && npm install && npm test`.

You can put your own extension in `testapp`'s `package.json` and it will work. I tested with your `dpd-passport` and saw no issues.

**Comment by Maddis1337** [comment]
@andreialecu then this sounds more or less like the solution I came up with. You can view it here in action:
https://bitbucket.org/simpletechs/dpd-passport/src/bd6f76925ed270b7bc4dcdd01138e4c3250a5f32/?at=develop

**Comment by andreialecu** [comment]
Yup. I saw your approach. The difference is that this uses a headless browser and /dpd.js, which may be more important when testing extensions that end users will work with.

**Comment by coderofsalvation** [comment]
Hi, might not be totally related, but i created a drop-in plugin to easily do e2e endpoint testing:

https://github.com/coderofsalvation/dpd-test

I needed this to test permissions/roles on a buildserver.
